### ISSUE TITLE
Empty Intervals - numElements and intersection

### ISSUE THREAD

**Issue by bogovicj** [imglib/imglib2#300]
I'd like to detect when the intersection of two intervals is empty, and I came across some surprising behavior.

```
// make an interval
FinalInterval A = new FinalInterval( 
		new long[]{ 0, 0, 0 },
		new long[]{ 5, 5, 5 });

// make B such that A intersect B is empty
FinalInterval B = new FinalInterval( 
		new long[]{ 10, 0, 0 }, 
		new long[] { 15, 5, 5 });

// compute the intersection
FinalInterval AandB = Intervals.intersect( A, B );
System.out.println( AandB );

// how many elements in the intersection
System.out.println( Intervals.numElements( AandB ));
```

The output is:
```
FinalInterval [(10, 0, 0) -- (5, 5, 5) = -4x6x6]
-144
```

I was surprised to see `Intervals.numElements` return a negative number.

If the convention is "anything with zero or fewer elements is empty" that's fine, but this output is still surprising, which is why I'd prefer `numElements` return always a non-negative number.  

As well, could we add to the `Interval` documentation something like: 
> An `Interval` for which `max(d)` < `min(d)` for any d, should be treated as empty.

**Comment by bogovicj** [comment]
> "anything with zero or fewer elements is empty"

This does not work with the current intersection implementation because:
```
FinalInterval A = new FinalInterval( 
		new long[]{ 0, 0, 0 },
		new long[]{ 5, 5, 5 });

// make B such that A intersect B is empty
FinalInterval B = new FinalInterval( 
		new long[]{ 10, 0, 10 }, 
		new long[] { 5, 5, 5 });

// compute the intersection
FinalInterval AandB = Intervals.intersect( A, B );
System.out.println( Intervals.numElements( AandB ));
```
returns `96`

i.e., the number of elements of the intersection of two non-intersecting intervals currently can return a positive number.

**Comment by imagejan** [comment]
I think the misconception here is that you call the constructor of `B` with a `min` that is larger than `max` in (at least) one dimension.

The following (Groovy) script works as expected for me:

```groovy
import net.imglib2.FinalInterval
import net.imglib2.util.Intervals

// make an interval
FinalInterval A = new FinalInterval( 
		new long[]{ 0, 0, 0 },
		new long[]{ 5, 5, 5 });

// make B such that A intersect B is empty
FinalInterval B = new FinalInterval( 
		new long[]{ 6, 0, 0 }, 
		new long[] { 10, 5, 5 });

// compute the intersection
FinalInterval AandB = Intervals.intersect( A, B );
println( AandB );

// how many elements in the intersection
println( Intervals.numElements( AandB ));
```

**Comment by bogovicj** [comment]
Indeed, that was a typo - but the issue persists when A and B are defined correctly.
(I edited) Sorry for confusion

**Comment by bogovicj** [comment]
* [`Intervals.isEmpty` ](https://github.com/imglib/imglib2/blob/master/src/main/java/net/imglib2/util/Intervals.java#L648) exists and behaves as expected

Here are two proposed chages, and one "do nothing" option. I spoke with @axtimwalde not long ago and we came up with Option C

## Option A

* Do nothing, because I should have used `Intervals.isEmpty`

## Option B

The draft PR above takes this approach.

* Adopt the convention: An Interval for which `max(d) < min(d)` for any d, should be treated as empty.
* `Intervals.intersection` can remain untouched
* `Intervals.numElements` changes
   * now checks if any max(d) < min(d) for any d and returns 0 if so

## Option C

* Adopt the convention: An Interval for which `max(d) < min(d)` is invalid. 
   * i.e. don't ever specifically check for this case
* `Intervals.intersection` changes
   * returns a new class `EmptyInterval` (see below) if the interval is empty
* `Intervals.numElements` can remain untouched

### `EmptyInterval`
* An implementation of `Interval` such that  
   * `Intervals.numElements( EmptyInterval)`  returns 0
   * `Intervals.intersection( interval EmptyInterval)`  returns `EmptyInterval`
   * `Intervals.union( interval, EmptyInterval)`  returns `Interval`

Still would have to consider how other methods behave, such as: `Intervals.contains`, `Intervals.expand` etc... maybe it's not such a great idea after all

**Comment by tpietzsch** [comment]
I would vote for **option B**.
And we would also need to fix `Intervals.union()` for that (as it is, the union of two empty intervals may become non-empty).

If we choose another option, we should also adopt it in other places, e.g., https://github.com/imglib/imglib2-roi/blob/f1e5daae75cab827790293b0bd9c9859f160d7c8/src/main/java/net/imglib2/roi/Bounds.java#L362-L403

**Comment by hanslovsky** [comment]
> And we would also need to fix `Intervals.union()` for that (as it is, the union of two empty intervals may become non-empty).

The issue is that, mathematically, a union of intervals is not necessarily an interval. The mathematically correct solution would be to introduce a base interface `Set` and have `Intervals.union()` return a `Set` but I do not think this is practical. I think we need to come up with a clearly defined way for what the union of two `Interval`s means in ImgLib2 terms. Something like
```
Return
    an empty `Interval` if there is no overlap,
    the smallest `Interval` that contains both `Interval`s otherwise
```

**Comment by axtimwalde** [comment]
The union of two `Intervals` according to our API is defined as an `Interval` with `min(a.min, b.min)` and `max(a.max, b.max)`. We can implement fancier unions elsewhere.  This union works for empty intervals in that it can create non-empty intervals. We can also add tests for `min < max` and thereby make all `Intervals` valid?

**Comment by tpietzsch** [comment]
I would add special treatment for empty intervals (with max<min) to `Intervals.union()`: The union of an interval `A` and an empty interval `B` returns `A`. I think it would be less confusing.
We could add another method that has the current behaviour. Ot the other way around, but I don't know what would be a good method name.

**Comment by tischi** [comment]
@bogovicj 

There are serious performance issues in https://github.com/imglib/imglib2-roi, making some functionality basically unusable when recursively joining more than ~10 intervals/rois, see: https://github.com/imglib/imglib2-roi/issues/60

I think I have now some ideas how to fix this, but in order to actually do it I need to know what our implementations for `Intervals.union()` and `Intervals.intersect()` and `Intervals.isEmpty()` will be, because I think we should make that consistent with the behavior of the equivalent classes `IntersectionRealInterval` and `UnionRealInterval`, which are the classes that suffer the performance issues.

In a nutshell: is there a chance to wrap things up here such that I could proceed with the other issue?

**Comment by bogovicj** [comment]
Hi @tischi 
This PR https://github.com/imglib/imglib2/pull/301 is ready to be merged from my point of view.

It is ["Option B" above  ](https://github.com/imglib/imglib2/issues/300#issuecomment-772141171).  `Intervals.numElements` changes (adding a check), but that's the only modification.  `union` and `intersection` are untouched.
